### PR TITLE
Ucx opts

### PR DESCRIPTION
**Pull request by JaeseungYeom**
imported Ian's branch PR #58 from TauferLab fork as Hari does not have access.

### CODE REVIEW

**Comment by JaeseungYeom** [comment]
I think we should breakdown this PR into multiple PRs.
There are many changes. We not only have to review these for merging but also be careful not to create conflicts with outstanding PRs.
Some changes are crucial. Some can be applied later. Some will be irrelevant.

- DTL related changes (crucial)
- There are modifications for caliper instrumentation. (later)
- Some modifications are formatting and minor changes that are not really crucial. (later if relevant) 
- We will also drop autoconfig-based build later as there will be complicated dependencies in the near future. (irrelevant)
- We will also remove dead code like urpc. (irrelevant)

It is best If a new PR can be created containing only those commits related dtl changes. Otherwise, I can give it a try to make this PR as minimal as possible for easy digestion. Let me know your opinion.

File: src/dyad/common/dyad_rc.h [hunk 1]
```diff
@@ -18,30 +20,36 @@ extern "C" {
...
+    DYAD_RC_UCXWAIT_FAIL = -15,     // UCX wait (either custom or
+                                    // 'ucp_worker_wait') failed
+    DYAD_RC_UCXEP_FAIL = -16,       // An operation on a ucp_ep_h failed
+    DYAD_RC_UCXCOMM_FAIL = -17,     // UCX communication routine failed
+    DYAD_RC_UCXMMAP_FAIL = -18,     // Failed to perform operations with ucp_mem_map
+    DYAD_RC_RPC_FINISHED = -19,     // The Flux RPC responded with ENODATA (i.e.,
+                                    // end of stream) sooner than expected
+    DYAD_RC_BAD_B64DECODE = -18,    // Decoding of data w/ base64 failed
+    DYAD_RC_BAD_COMM_MODE = -19,    // Invalid comm mode provided to DTL
+    DYAD_RC_UNTRACKED = -20,        // Provided path is not tracked by DYAD
```
Comment (JaeseungYeom) [inline src/dyad/common/dyad_rc.h:48]: The only changes with these lines are the indentation for the comment at the end of each line. (extra space for alignment)

Comment (JaeseungYeom) [reply src/dyad/common/dyad_rc.h:48]: Next 4 are new.

**Comment by hariharan-devarajan** [comment]
@JaeseungYeom we can only include optimization from your list and remove the rest that would be ideal. 

We can do instrumentation stuff later as I am also working on a way to do it cleanly with DLIO profiler

File: src/dyad/core/dyad_core.c [hunk 2]
```diff
@@ -511,10 +522,14 @@ DYAD_CORE_FUNC_MODS dyad_rc_t dyad_cons_store (const dyad_ctx_t* restrict ctx,
     rc = DYAD_RC_OK;
 
 pull_done:
+    DYAD_PERF_REGION_END (ctx->perf_handle, "dyad_write_data_for_consumer");
+    if (file_data != NULL) {
+        ctx->dtl_handle->return_buffer (ctx->dtl_handle, (void**)&file_data);
+    }
```
Comment (JaeseungYeom) [inline src/dyad/core/dyad_core.c:528]: file_data deallocation has been moved after "consume_done" in `dyad_consume()` in the latest code base.

File: src/dyad/dtl/CMakeLists.txt [hunk 3]
```diff
@@ -25,6 +24,19 @@ set(DTL_PUBLIC_HEADERS ${DTL_PUBLIC_HEADERS} ${FLUX_PUBLIC_HEADERS})
...
+#    set(DTL_PRIVATE_HEADERS ${DTL_PRIVATE_HEADERS} ${FLUX_PRIVATE_HEADERS})
+#    set(DTL_PUBLIC_HEADERS ${DTL_PUBLIC_HEADERS} ${FLUX_PUBLIC_HEADERS})
+#else()
+#    message(FATAL_ERROR "-- [${PROJECT_NAME}] Unknown DYAD_DATA_PLANE configuration ${DYAD_DATA_PLANE}")
+#endif()
+
+#add_library(${PROJECT_NAME}_dtl SHARED ${DTL_SRC})
+#target_link_libraries(${PROJECT_NAME}_dtl PUBLIC utils Jansson::Jansson flux::core)
+#>>>>>>> fbec6e7 (refactored and initial changes to build.)
+target_link_libraries(${PROJECT_NAME}_dtl PUBLIC ${PROJECT_NAME}_utils Jansson::Jansson flux::core ${PROJECT_NAME}_perf)
```
Comment (JaeseungYeom) [inline src/dyad/dtl/CMakeLists.txt:39]: I was not sure how to resolve this conflict while rebasing. So, I left it as a comment block.

File: src/dyad/dtl/CMakeLists.txt [hunk 4]
```diff
@@ -10,8 +9,8 @@ set(FLUX_PRIVATE_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/flux_dtl.h)
 set(FLUX_PUBLIC_HEADERS )
 
 # UCX implementation for DTL
-set(UCX_DTL_SRC ${CMAKE_CURRENT_SOURCE_DIR}/ucx_dtl.c)
-set(UCX_PRIVATE_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/ucx_dtl.h)
+set(UCX_DTL_SRC ${CMAKE_CURRENT_SOURCE_DIR}/ucx_dtl.c ${CMAKE_CURRENT_SOURCE_DIR}/ucx_ep_cache.cpp)
+set(UCX_PRIVATE_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/ucx_dtl.h ${CMAKE_CURRENT_SOURCE_DIR}/ucx_ep_cache.h)
```
Comment (JaeseungYeom) [inline src/dyad/dtl/CMakeLists.txt:13]: These lines are relevant to DTL.

File: src/dyad/core/dyad_core.c [hunk 5]
```diff
@@ -600,24 +628,26 @@ dyad_rc_t dyad_init (bool debug,
...
-        return DYAD_RC_NOCTX;
+        rc = DYAD_RC_NOCTX;
+        goto init_region_finish;
     }
     strncpy ((*ctx)->kvs_namespace, kvs_namespace, namespace_len + 1);
     // Initialize the DTL based on the value of dtl_mode
     // If an error occurs, log it and return an error
     FLUX_LOG_INFO ((*ctx)->h, "DYAD_CORE: inintializing DYAD DTL");
-    rc = dyad_dtl_init (&(*ctx)->dtl_handle, dtl_mode, (*ctx)->h, (*ctx)->debug);
+    rc = dyad_dtl_init (&(*ctx)->dtl_handle, dtl_mode, DYAD_COMM_RECV, (*ctx)->h, (*ctx)->debug, (*ctx)->perf_handle);
```
Comment (JaeseungYeom) [inline src/dyad/core/dyad_core.c:647]: This line is relevant to DTL.

File: src/dyad/dtl/dyad_dtl_impl.h [hunk 6]
```diff
@@ -41,17 +40,17 @@ typedef enum dyad_dtl_comm_mode dyad_dtl_comm_mode_t;
...
-                             const flux_msg_t* packed_obj,
-                             char** upath);
+    dyad_rc_t (*rpc_unpack) (struct dyad_dtl* self, const flux_msg_t* packed_obj, char** upath);
     dyad_rc_t (*rpc_respond) (struct dyad_dtl* self, const flux_msg_t* orig_msg);
     dyad_rc_t (*rpc_recv_response) (struct dyad_dtl* self, flux_future_t* f);
-    dyad_rc_t (*establish_connection) (struct dyad_dtl* self,
-                                       dyad_dtl_comm_mode_t comm_mode);
+    dyad_rc_t (*get_buffer) (struct dyad_dtl* self, size_t data_size, void** data_buf);
+    dyad_rc_t (*return_buffer) (struct dyad_dtl* self, void** data_buf);
+    dyad_rc_t (*establish_connection) (struct dyad_dtl* self);
```
Comment (JaeseungYeom) [inline src/dyad/dtl/dyad_dtl_impl.h:53]: These changes are relevant to DTL.

File: src/dyad/dtl/flux_dtl.c [hunk 7]
```diff
@@ -19,40 +24,57 @@ dyad_rc_t dyad_dtl_flux_init (dyad_dtl_t* self,
     self->rpc_unpack = dyad_dtl_flux_rpc_unpack;
     self->rpc_respond = dyad_dtl_flux_rpc_respond;
     self->rpc_recv_response = dyad_dtl_flux_rpc_recv_response;
+    self->get_buffer = dyad_dtl_flux_get_buffer;
+    self->return_buffer = dyad_dtl_flux_return_buffer;
```
Comment (JaeseungYeom) [inline src/dyad/dtl/flux_dtl.c:28]: These lines are relevant to DTL.

File: src/dyad/dtl/flux_dtl.c [hunk 8]
```diff
@@ -2,15 +2,20 @@
...
+    DYAD_PERF_REGION_BEGIN (self->perf_handle, "dyad_dtl_flux_init");
     self->private.flux_dtl_handle = malloc (sizeof (struct dyad_dtl_flux));
     if (self->private.flux_dtl_handle == NULL) {
         FLUX_LOG_ERR (h, "Cannot allocate the Flux DTL handle\n");
-        return DYAD_RC_SYSFAIL;
+        rc = DYAD_RC_SYSFAIL;
+        goto dtl_flux_init_region_finish;
     }
     self->private.flux_dtl_handle->h = h;
+    self->private.flux_dtl_handle->comm_mode = comm_mode;
```
Comment (JaeseungYeom) [inline src/dyad/dtl/flux_dtl.c:18]: This line is relevant to DTL.

File: src/dyad/dtl/flux_dtl.c [hunk 9]
```diff
@@ -66,15 +88,48 @@ dyad_rc_t dyad_dtl_flux_rpc_recv_response (dyad_dtl_t* self, flux_future_t* f)
...
+        goto flux_ret_buf_done;
+    }
+    free (*data_buf);
+    rc = DYAD_RC_OK;
+
+flux_ret_buf_done:
+    return rc;
+}
+
+dyad_rc_t dyad_dtl_flux_establish_connection (dyad_dtl_t* self)
```
Comment (JaeseungYeom) [inline src/dyad/dtl/flux_dtl.c:123]: These lines are relevant to DTL.

File: src/dyad/dtl/flux_dtl.c [hunk 10]
```diff
@@ -118,18 +179,18 @@ dyad_rc_t dyad_dtl_flux_recv (dyad_dtl_t* self, void** buf, size_t* buflen)
...
-        if (*buf == NULL) {
-            FLUX_LOG_ERR (dtl_handle->h, "Could not allocate space for file buffer");
-            dyad_rc = DYAD_RC_SYSFAIL;
-            goto finish_recv;
-        }
+    dyad_rc = self->get_buffer (self, *buflen, buf);
+    if (DYAD_IS_ERROR (dyad_rc)) {
+        *buf = NULL;
+        *buflen = 0;
+        goto finish_recv;
```
Comment (JaeseungYeom) [inline src/dyad/dtl/flux_dtl.c:186]: The lines are relevant to DTL.

File: src/dyad/dtl/flux_dtl.c [hunk 11]
```diff
@@ -118,18 +179,18 @@ dyad_rc_t dyad_dtl_flux_recv (dyad_dtl_t* self, void** buf, size_t* buflen)
...
     }
     memcpy (*buf, tmp_buf, tmp_buflen);
     dyad_rc = DYAD_RC_OK;
 finish_recv:
-    flux_future_reset (dtl_handle->f);
+    if (dtl_handle->f != NULL)
+        flux_future_reset (dtl_handle->f);
```
Comment (JaeseungYeom) [inline src/dyad/dtl/flux_dtl.c:192]: The lines are relevant to DTL.

File: src/dyad/dtl/flux_dtl.c [hunk 12]
```diff
@@ -144,12 +205,16 @@ dyad_rc_t dyad_dtl_flux_close_connection (dyad_dtl_t* self)
 
 dyad_rc_t dyad_dtl_flux_finalize (dyad_dtl_t** self)
 {
-    if (self == NULL || *self == NULL)
+    if (self == NULL || *self == NULL) {
         return DYAD_RC_OK;
+    }
+    dyad_perf_t* perf_handle = (*self)->perf_handle;
```
Comment (JaeseungYeom) [inline src/dyad/dtl/flux_dtl.c:211]: The lines are relevant to DTL.

File: src/dyad/modules/dyad.c [hunk 13]
```diff
@@ -70,6 +70,11 @@ static void freectx (void *arg)
         dyad_dtl_finalize (&(ctx->dtl_handle));
         ctx->dtl_handle = NULL;
     }
+    if (ctx->perf_handle != NULL) {
+        ctx->perf_handle->flush (ctx->perf_handle);
+        dyad_perf_finalize (&(ctx->perf_handle));
+        ctx->perf_handle = NULL;
+    }
```
Comment (JaeseungYeom) [inline src/dyad/modules/dyad.c:77]: These lines are relevant to DTL.

File: src/dyad/modules/dyad.c [hunk 14]
```diff
@@ -158,21 +178,32 @@ dyad_fetch_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t *msg,
...
     }
-    if ((inlen = read_all (fd, &inbuf)) < 0) {
+    FLUX_LOG_INFO (h, "Getting file size");
+    file_size = get_file_size (fd);
+    FLUX_LOG_INFO (h, "Getting buffer for reading the file");
+    rc = ctx->dtl_handle->get_buffer (ctx->dtl_handle, file_size, &inbuf);
+    FLUX_LOG_INFO (h, "Reading file data into buffer");
+    if ((inlen = read_all (fd, inbuf, file_size)) < 0) {
         FLUX_LOG_ERR (h, "DYAD_MOD: Failed to load file \"%s\".\n", fullpath);
+        close (fd);
```
Comment (JaeseungYeom) [inline src/dyad/modules/dyad.c:196]: These are relevant to DTL, but use less LOG_INFO.

File: src/dyad/modules/dyad.c [hunk 15]
```diff
@@ -158,21 +178,32 @@ dyad_fetch_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t *msg,
...
         goto fetch_error;
     }
+    FLUX_LOG_INFO (h, "Closing file pointer");
     close (fd);
+    DYAD_PERF_REGION_END (ctx->perf_handle, "dyad_server_read_data");
     FLUX_LOG_INFO (h, "Is inbuf NULL? -> %i\n", (int)(inbuf == NULL));
 
     FLUX_LOG_INFO (h, "Establish DTL connection with consumer");
-    rc = ctx->dtl_handle->establish_connection (ctx->dtl_handle, DYAD_COMM_SEND);
+    rc = ctx->dtl_handle->establish_connection (ctx->dtl_handle);
```
Comment (JaeseungYeom) [inline src/dyad/modules/dyad.c:206]: This is relevant to DTL.

File: src/dyad/modules/dyad.c [hunk 16]
```diff
@@ -182,125 +213,205 @@ dyad_fetch_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t *msg,
     rc = ctx->dtl_handle->send (ctx->dtl_handle, inbuf, inlen);
     FLUX_LOG_INFO (h, "Close DTL connection with consumer");
     ctx->dtl_handle->close_connection (ctx->dtl_handle);
-    free (inbuf);
+    ctx->dtl_handle->return_buffer (ctx->dtl_handle, &inbuf);
```
Comment (JaeseungYeom) [inline src/dyad/modules/dyad.c:216]: This is relevant to DTL.

File: src/dyad/modules/dyad.c [hunk 17]
```diff
@@ -182,125 +213,205 @@ dyad_fetch_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t *msg,
...
-
+    rc = dyad_perf_init (&(ctx->perf_handle), true, opts);
+    if (DYAD_IS_ERROR (rc))
+        goto open_done;
+    rc = dyad_dtl_init (&(ctx->dtl_handle),
+                        dtl_mode,
+                        DYAD_COMM_SEND,
+                        h,
+                        ctx->debug,
+                        ctx->perf_handle);
```
Comment (JaeseungYeom) [inline src/dyad/modules/dyad.c:263]: This is relevant to DTL except the last argument. The only thing added is the third argument, DYAD_COMM_SEND.

**Comment by JaeseungYeom** [comment]
The below is the list of the changes I think need to be merged in as the first PR.

src/dyad/dtl/ucx_ep_cache.h
src/dyad/dtl/ucx_ep_cache.cpp
src/dyad/dtl/ucx_dtl.h
src/dyad/dtl/ucx_ep_cache.h
src/dyad/dtl/ucx_ep_cache.c
src/dyad/dtl/flux_dtl.h
src/dyad/dtl/ucx_dtl.c
src/dyad/dtl/ucx_dtl.h


src/dyad/common/dyad_rc.h
  lines 49-52
src/dyad/core/dyad_core.c
  lines 405, 433,  526-528 might be a conflict, 647

src/dyad/dtl/CMakeLists.txt
  lines 12-13

src/dyad/dtl/dyad_dtl_impl.h
  lines 51-53

src/dyad/dtl/flux_dtl.c
  lines 18, 27, 28, 91-123, 182-186, 191, 192, 208-211

src/dyad/modules/dyad.c
 lines 74-77, 90-96, 206, 216, 258-262

**Comment by JaeseungYeom** [comment]
I think the best way now is to merge PR #65 first by @hariharan-devarajan 
Then, @ilumsden can create a new PR based on the latest code with the changes listed above.

**Pull request closed without merging**